contracts/pcv/sentinel/PCVSentinel.sol

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "../../refs/CoreRef.sol";
+import "./IPCVSentinel.sol";
+import "../IPCVDeposit.sol";
+import "../../libs/CoreRefPauseableLib.sol";
+import "./IGuard.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+/**
+ * @title PCV Sentinel
+ * @dev the PCV Sentinel should be granted the role Guardian
+ * @notice The PCV Sentinel is a automated extension of the Guardian.
+ * The Guardian can add Guards to the PCV Sentinel. Guards run checks
+ * and provide addresses and calldata for the Sentinel to run, if needed.
+ */
+contract PCVSentinel is IPCVSentinel, CoreRef {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    // The set of all guards
+    EnumerableSet.AddressSet private guards;
+
+    constructor(
+        address _core
+    ) CoreRef(_core) {
+        _setContractAdminRole(keccak256("PCV_SENTINEL_ADMIN_ROLE"));
+    }
+
+    // ---------- Read-Only API ----------
+
+    /**
+     * @notice returns whether or not the given address is a guard
+     * @param guard the address to check
+     */
+    function isGuard(address guard) 
+        external 
+        view
+        override
+        returns (bool) 
+    {
+        return guards.contains(guard);
+    }
+
+    /**
+     * @notice returns a list of all guards
+     */
+    function allGuards() 
+        external
+        view
+        override
+        returns (address[] memory all) 
+    {
+        all = new address[](guards.length());
+
+        for (uint i = 0; i < guards.length(); i++) {
+            all[i] = guards.at(i);
+        }
+
+        return all;
+    }
+
+    /**
+     * @notice returns the list of all guards that pass their check
+     * @dev you *can* use this on-chain if you want, but this is made
+     * to be called off-chain to pass into protecManys and not hyper-efficient
+     */
+    function checkAll()
+        external 
+        view
+        override 
+        returns (address[] memory) 
+    {
+        uint256 numGuardsCanProtec;
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
+                numGuardsCanProtec++;
+            }
+        }
+
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
```

Comment:
you could cache these check() results in a boolean array to save external calls (I know it's not supposed to be called onchain, but...)

Reply:
yeah if I want to use this onchain I'll definitely make an optimized version of this, makes it harder to read & audit though

Reply:
wait nvm this isn't hard. good callout.